### PR TITLE
close #830 edit line item fields backend

### LINKED ISSUE #830: Edit line item duration and other fields

### PR DESCRIPTION
**Pull request by Thornsopanha**
# Edit Screen UI
Screenshot 2023-11-29 at 10 57 55 AM

### CODE REVIEW

File: app/controllers/spree/admin/line_items_controller.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        @order = @line_item.order
+        if @line_item.update(line_item_params)
+          redirect_to admin_orders_path(@line_item)
+        else
+          render :edit
+        end
+      end
+
+      private
+
```
Comment (theachoem) [inline app/controllers/spree/admin/line_items_controller.rb:27]:
```
def location_after_save
  admin_cart_orders_path(@order)
end
```

**Review by theachoem — APPROVED** [review]
LGTM

File: config/routes.rb [hunk 2]
```diff
@@ -101,6 +101,8 @@
         put :fire_notification
         put :queue_webhooks_requests
       end
+
+      resources :line_items
```
Comment (theachoem) [inline config/routes.rb:105]: resources :line_items, only: %i[update edit]

File: app/views/spree/admin/line_items/edit.html.erb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<%= render partial: 'spree/admin/shared/order_tabs', locals: { current: :cart } %>
+<%= render partial: 'spree/admin/shared/error_messages', locals: { target: @line_item } %>
+<%= form_with model: @line_item, url: { action: 'update' } do |f| %>
+<%= render partial: 'form', locals: { f: f } %>
+  <div data-hook="admin_line_item_edit_form_button">
+    <%= render partial: 'spree/admin/shared/edit_resource_links', locals: { collection_url: spree.admin_orders_path } %>
+  </div>
+<% end %>
```
Comment (theachoem) [inline app/views/spree/admin/line_items/edit.html.erb:8]:
```rb
<%= render partial: 'spree/admin/shared/order_tabs', locals: { current: :cart } %>
<%= render partial: 'spree/admin/shared/error_messages', locals: { target: @line_item } %>
<%= form_with model: @line_item, url: { action: 'update' } do |f| %>
  <%= render partial: 'form', locals: { f: f } %>
  <div data-hook="admin_line_item_edit_form_button">
    <%= render partial: 'spree/admin/shared/edit_resource_links', locals: { collection_url: spree.admin_orders_path } %>
  </div>
<% end %>
```

**Review by theachoem — APPROVED** [review]
LGTM

File: app/controllers/spree/admin/line_items_controller.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      def permitted_resource_params
+        params.require(:line_item).permit(:quantity, :from_date, :to_date)
+      end
+
+      def update_service
+        Spree::Dependencies.line_item_update_service.constantize
+      end
+
+      def load_data
+        @line_item = LineItem.find(params[:id])
```
Comment (channainfo) [inline app/controllers/spree/admin/line_items_controller.rb:34]: let's find the line item from the order to make sure that the route is nested under the order resources and the line_item work only if the order is valid:

```
@order = Spree::Order.find(params[:order_id])
@line_item = @order.line_items.find(params[:id])
```

**Review by channainfo — CHANGES_REQUESTED** [review]

**Pull request merged**